doc/sonic-sub-port-intf-hld.md

```diff
@@ -0,0 +1,483 @@
+
+sub_intf_attrs[0].id = SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID;
+sub_intf_attrs[0].value.oid = vrf_oid;
+
+sub_intf_attrs[1].id = SAI_ROUTER_INTERFACE_ATTR_TYPE;
+sub_intf_attrs[1].value.s32 = SAI_ROUTER_INTERFACE_TYPE_SUB_PORT;
+
+sub_intf_attrs[2].id = SAI_ROUTER_INTERFACE_ATTR_PORT_ID;
+sub_intf_attrs[2].value.oid = parent_port_oid;  /* oid of the parent physical port or port channel */
+
+sub_intf_attrs[3].id = SAI_ROUTER_INTERFACE_ATTR_VLAN_ID;
+sub_intf_attrs[3].value.oid = vlan_oid;
+
+sai_mac_t mac = {0x00, 0xe0, 0xec, 0xc2, 0xad, 0xf1};
+sub_intf_attrs[4].id = SAI_ROUTER_INTERFACE_ATTR_SRC_MAC_ADDRESS;
+memcpy(sub_intf_attrs[4].value.mac, mac, sizeof(sai_mac_t));
+
+sub_intf_attrs[5].id = SAI_ROUTER_INTERFACE_ATTR_MTU;
+sub_intf_attrs[5].value.u32 = 9100;
+
+uint32_t sub_intf_attrs_count = 6;
+sai_status_t status = create_router_interface(&rif_id, switch_oid, sub_intf_attrs_count, sub_intf_attrs);
+```
+
+### 2.4.2 Set a sub port interface mtu
+```
+sai_attribute_t sub_intf_attr;
+sub_intf_attr.id = SAI_ROUTER_INTERFACE_ATTR_MTU;
+sub_intf_attr.value.u32 = 9000;
+
+sai_status_t status = set_router_interface_attribute(rif_id, &attr);
+```
+
+### 2.4.3 Remove a sub port interface
+```
+sai_status_t status = remove_router_interface(rif_id);
+```
+
+## 2.5 Linux integration
+### 2.5.1 Host sub port interfaces
+
+We use iproute2 package to manage host sub port interfaces.
+Specifically, we use `ip link add link <parent_port_name> name <subif_name> type vlan id <vlan_id>` to create a host sub port interface.
+This command implies the dependancy that a parent host interface must be created before the creation of a host sub port interface.
+At creation, a host sub port interface is always set to admin status up.
+
+Example:
+```
+ip link add link Ethernet64 name Ethernet64.10 type vlan id 10
+ip link set Ethernet64.10 mtu 9100
+ip link set Ethernet64.10 up
+```
+```
+ip link del Ethernet64.10
+```
+
+We use `ip address` and `ip -6 address` to add and remove ip adresses on a host sub port interface.
+
+Example:
+```
+ip address add 192.168.0.1/24 dev Ethernet64.10
+```
+
+### 2.5.2 Route, neighbor subsystems
+
+Once the host sub port interfaces are properly set, route and neighbor subsystems should function properly on sub port interfaces.
+fpmsyncd should receive route add/del updates on sub port interfaces from zebra over TCP socket port 2620. These updates are received in the format of netlink messages.
+neighsyncd should receive neigh add/del netlink messages on sub port interfaces from its subscription to the neighbor event notification group RTNLGRP_NEIGH.
+
+Internally, a sub port interface is represented as a Port object to be perceived seamlessly by NeighOrch and RouteOrch to create neighor and route entries, respectively, on it.
+
+
+# 3 Event flow diagrams
+## 3.1 Sub port interface creation
+![](https://github.com/wendani/SONiC/blob/sub_port_master/images/sub_interface_hld/sub_intf_creation_flow.png)
+
```

Comment:
In the section title 'SCOPE', it is mentioned that each sub port interface is considered to stay in a separate bridge domain and is useful. Given this, Is vlan-10 object creation by intfOrch necessary? If yes, it needs to be mentioned how 

1)  vlan-10 will be re-used when another support with vlan-10 is created 

2) when will vlan-10 be deleted.

In the current call sequence in Sec 3.3, vlan-10 object is lingering around after the subport has been deleted.

I think this complication arises from the way SAI_ROUTER_INTERFACE_ATTR_VLAN_ID is defined in the SAI spec. Since the VLANs mentioned in the support configuration have only port namespace, we really don't need a global VLAN object and just a uint16_t type should be sufficient.

    /**
     * @brief Associated Vlan
     *
     * @type sai_object_id_t <<<<<<<<
     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
     * @objects SAI_OBJECT_TYPE_VLAN <<<<<<<
     * @condition SAI_ROUTER_INTERFACE_ATTR_TYPE == SAI_ROUTER_INTERFACE_TYPE_VLAN or SAI_ROUTER_INTERFACE_ATTR_TYPE == SAI_ROUTER_INTERFACE_TYPE_SUB_PORT
     */
    SAI_ROUTER_INTERFACE_ATTR_VLAN_ID,